File: src/main/java/org/cubewhy/celestial/files/ConfigFile.java
package org.cubewhy.celestial.files;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import java.io.*;

@Slf4j
public class ConfigFile {
    @Getter
    private JsonObject config;
    public final File file;

    public ConfigFile(File file) {
        this.file = file;
        this.load();
    }

    public ConfigFile setValue(String key, String value) {
        this.config.addProperty(key, value);
        return this.save();
    }

    public ConfigFile setValue(String key, char value) {
        this.config.addProperty(key, value);
        return this.save();
    }

    public ConfigFile setValue(String key, int value) {
        this.config.addProperty(key, value);
        return this.save();
    }

    public ConfigFile setValue(String key, boolean value) {
        this.config.addProperty(key, value);
        return this.save();
    }

    public ConfigFile setValue(String key, JsonObject value) {
        this.config.add(key, value);
        return this.save();
    }

    public ConfigFile initValue(String key, JsonElement value) {
        if (!this.config.has(key)) {
            log.info("Init value " + key + " -> " + value);
            this.config.add(key, value);
        }
        return this.save();
    }

    public ConfigFile initValue(String key, String value) {
        return this.initValue(key, new JsonPrimitive(value));
    }

    public ConfigFile initValue(String key, int value) {
        return this.initValue(key, new JsonPrimitive(value));
    }

    public JsonElement getValue(String key) {
        return this.config.get(key);
    }

    public ConfigFile save() {
        try {
            BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(this.file));
            bufferedWriter.write(config.toString());
            bufferedWriter.flush();
            bufferedWriter.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return this;
    }

    public ConfigFile load() {
        Gson gson = new Gson();
        BufferedReader bufferedReader;
        boolean successful = false;

        while (!successful) {
            try {
                bufferedReader = new BufferedReader(new FileReader(this.file));
                config = gson.fromJson(bufferedReader, JsonObject.class);
                if (config == null) {
                    config = new JsonObject();
                }
                successful = true;
            } catch (FileNotFoundException e) {

                try {
                    if (!this.file.getParentFile().exists()) {
                        this.file.getParentFile().mkdirs();
                    }
                    this.file.createNewFile();
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
            }
        }
        return this;
    }

}

File: src/main/kotlin/org/cubewhy/celestial/files/ConfigFile.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.files

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.*
import java.nio.charset.StandardCharsets

open class ConfigFile(val file: File) {
    lateinit var config: JsonObject
        private set

    init {
        this.load()
    }

    fun setValue(key: String, value: String): ConfigFile {
        config.addProperty(key, value)
        return this
    }

    fun setValue(key: String, value: Char): ConfigFile {
        config.addProperty(key, value)
        return this
    }

    fun setValue(key: String, value: Number): ConfigFile {
        config.addProperty(key, value)
        return this
    }

    fun setValue(key: String, value: Boolean): ConfigFile {
        config.addProperty(key, value)
        return this
    }

    fun setValue(key: String, value: JsonObject?): ConfigFile {
        config.add(key, value)
        return this
    }

    fun initValue(key: String, value: JsonElement?): ConfigFile {
        if (!config.has(key)) {
            log.info("Init value $key -> $value")
            config.add(key, value)
        }
        return this
    }

    fun initValue(key: String, value: String): ConfigFile {
        return this.initValue(key, JsonPrimitive(value))
    }

    fun initValue(key: String, value: Int): ConfigFile {
        return this.initValue(key, JsonPrimitive(value))
    }

    fun getValue(key: String): JsonElement {
        return config[key]
    }

    fun save(): ConfigFile {
        try {
            val bufferedWriter = BufferedWriter(FileWriter(this.file))
            bufferedWriter.write(config.toString())
            bufferedWriter.flush()
            bufferedWriter.close()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return this
    }

    private fun load(): ConfigFile {
        val gson = Gson()
        var bufferedReader: BufferedReader
        var successful = false

        while (!successful) {
            try {
                bufferedReader = BufferedReader(FileReader(this.file, StandardCharsets.UTF_8))
                config = with(gson.fromJson(bufferedReader, JsonObject::class.java)) {
                    this ?: JsonObject()
                }
                successful = true
            } catch (e: FileNotFoundException) {
                try {
                    if (!file.parentFile.exists()) {
                        file.parentFile.mkdirs()
                    }
                    file.createNewFile()
                } catch (ex: IOException) {
                    throw RuntimeException(ex)
                }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        return this
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(ConfigFile::class.java)
    }
}
